### PR TITLE
feat: add previous responses as a default

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Description
Use a previous response/default value as value to another variable into toml file

## Example

The following toml file describes the feature included in this PR, basically will be possible to use a variable template to get already assigned values from other variables. eg will be assigned to the `manifest` variable the value `my_project-other_project-manifest.md` if the user leave the all the inputs empty, however if the user populate the variable `project_one` with `tree` and variable `project_two` with `leaf` and leave `manifest` variable empty then the value assigned to manifest will be `tree-leaf-manifest.md`.

```toml
[[variables]]
name = "project_one"
default = "my_project"
prompt = "What's the name of your first project?"

[[variables]]
name = "project_two"
default = "other_project"
prompt = "What's the name of your second project?"

[[variables]]
name = "manifest"
default = "{{project_one}}-{{project_two}}-manifest.md"
prompt = "What's the manifest name file?"
```

The prompt default value is updated too!
image

## Issue
https://github.com/Keats/kickstart/issues/22

### CODE REVIEW

File: examples/default-as-variable/template.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+name = "Super basic"
```
Comment (Keats) [inline examples/default-as-variable/template.toml:1]: `default_from_variable` would be a better name for the file

File: src/definition.rs [hunk 2]
```diff
@@ -134,6 +143,38 @@ impl TemplateDefinition {
...
+                _ => cloned_str
+            }
+        }
+    }
+
+    cloned_str
+}
+
+fn has_template_variables<'a>(s: &'a String, _already_vars: &HashMap<String, Value>) -> Option<HashSet<&'a str>> {
+    let re = Regex::new(r"\{\{(?:[a-zA-Z][0-9a-zA-Z_]*)\}\}").unwrap();
```
Comment (Keats) [inline src/definition.rs:166]: that's going to compile that regex every time this function is called, you can lift it in some lazy static or once_cell, whatever this crate is using (or add one of these libs to it)

File: src/definition.rs [hunk 3]
```diff
@@ -113,11 +114,19 @@ impl TemplateDefinition {
                     continue;
                 }
                 Value::String(s) => {
+                    let variables = has_template_variables(&s, &vals);
+                    let default_value = match variables {
+                        Some(variables) => replace_with_previous_responses(
+                            variables, &vals, &s),
+                        None => s.clone(),
+                    };
```
Comment (Keats) [inline src/definition.rs:122]: I think doing it that way is more complicated than it actually requires.
A default value will always be a valid Tera template as it's just a string so you can just check whether `{{` and `}}` are in the string and handle both cases:

1. no: just use that string a default, no need to do anything
2. yes: render that string as a one off template with the variables so far added to the Tera context.

**Review by Keats — CHANGES_REQUESTED** [review]

**Comment by EclesioMeloJunior** [comment]
got it! never used Tera before, I will address your comments asap!

File: src/definition.rs [hunk 4]
```diff
@@ -114,13 +117,25 @@ impl TemplateDefinition {
                     continue;
                 }
                 Value::String(s) => {
-                    let variables = has_template_variables(&s, &vals);
-                    let default_value = match variables {
-                        Some(variables) => replace_with_previous_responses(
-                            variables, &vals, &s),
-                        None => s.clone(),
+                    let contains_template = has_template_variables(&s);
```
Comment (Keats) [inline src/definition.rs:120]: you can just check `if s.contains("{{") && s.contains("}}")` rather than using regex

File: src/utils.rs [hunk 5]
```diff
@@ -47,6 +48,18 @@ pub fn get_source(input: &str) -> Source {
     }
 }
 
+pub fn render_one_off_template(
+    content: &str,
+    context: &Context,
+    path: Option<PathBuf>,
+) -> Result<String> {
+    let mut tera = Tera::default();
```
Comment (Keats) [inline src/utils.rs:56]: there's actually a better way to do that now: https://docs.rs/tera/latest/tera/struct.Tera.html#method.one_off

**Review by Keats — CHANGES_REQUESTED** [review]

File: src/utils.rs (hunk 5 above)
Comment (EclesioMeloJunior) [reply src/utils.rs:56]: question: should we set the `autoscape` to `true` or transform it into a function parameter?

Comment (Keats) [reply src/utils.rs]: no autoescape needed there

Comment (EclesioMeloJunior) [reply src/utils.rs]: @Keats autoscape set to `false`!

**Comment by Keats** [comment]
Thanks, I'll merge it and tweak a bit later this week

**Review by Keats — APPROVED** [review]

**Pull request merged**